### ISSUE TITLE
[Epic #306] Phase 3: large-net throughput benchmarks + BURN_BACKENDS.md re-measurement

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#328]
**Epic**: #306 — Atari (ALE) environment + CNN policy
**Phase**: 3 of 4
**Phase Goal**: First honest CPU-vs-wgpu-vs-cuda crossover measurement on a large-net workload.

## Description

Per docs/BURN_BACKENDS.md, CPU currently wins 4.4–9.5× at small-net sizes, and the Nature-DQN CNN is the designated re-evaluation trigger:

- Add a throughput benchmark group for the CNN policy (#327) to `benches/trainer_throughput.rs` (forward + backward at Nature-DQN scale, batch sizes representative of DQN/PPO training).
- Re-measure CPU vs wgpu vs cuda on this workload (cuda numbers on alc-2 or equivalent; document the host).
- Update `docs/BURN_BACKENDS.md` with the large-net numbers and the crossover verdict (epic acceptance criterion #3).

## Acceptance Criteria

- [ ] Benchmark group in `benches/trainer_throughput.rs` covering the Nature-DQN CNN at training-representative batch sizes
- [ ] `docs/BURN_BACKENDS.md` updated with measured large-net CPU/wgpu/cuda numbers and the crossover conclusion
- [ ] Benches compile in CI; tests + clippy -D warnings + fmt + doc build clean

## Dependencies

Part of Epic #306. Blocked by #327 (CNN policy). GPU (cuda) numbers may require operator access to alc-2 — if so, land the bench + CPU/wgpu numbers and flag the cuda column for operator follow-up.

---
*Decomposed from Epic #306*

**Comment by rjwalters** [comment]
**Phase 1 spike clarification (docs/ALE_BINDING_STRATEGY.md, #324)**

Scope confirmed. Because Phase 2 binds ALE via a subprocess (Option D), benchmark the **CNN policy in isolation** — forward/backward on synthetic 84×84×4 batches — decoupled from env stepping, so the CPU/wgpu/cuda crossover measurement in \`docs/BURN_BACKENDS.md\` reflects the *network* and is not contaminated by subprocess IPC overhead.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating per user request (issue has `loom:architect` + `loom:epic-phase`). Grounded against `origin/main` after PR #332 merged the Nature-DQN CNN policy.

---

### Dependency status

`src/policy/atari_cnn.rs` is confirmed present on `origin/main` (verified via `git ls-tree`). The PR #332 dependency is satisfied. No `loom:blocked` needed.

---

### Benchmark harness conventions (builder must match these exactly)

`benches/trainer_throughput.rs` uses **criterion** with a fully generic `register_all<B: AutodiffBackend>(c: &mut Criterion, device: &B::Device, suffix: &str)` function that runs every group once per backend, tagging criterion group names as `{logical_group}/{suffix}` (e.g. `a2c_train_step/ndarray`, `a2c_train_step/wgpu`).

Key conventions:
- Each logical group maps to one `bench_<name><B>` free function taking `(c, device, suffix)`.
- Within each group there is typically one or two `bench_function` entries; existing names are `"synthetic_batch"`, `"replay_minibatch"`, or `"rollout_plus_update"`. For the CNN groups, use `"b32"` and `"b256"` to name the two batch-size variants.
- `Throughput::Elements` is set to `batch_size as u64` so criterion reports samples/sec alongside wall-clock.
- `iter_batched(..., BatchSize::SmallInput)` — the setup closure (untimed) builds the module and input tensor; the timed closure calls `black_box(result)`.
- GPU backends are registered at the `benches()` driver function, after the CPU registration, behind `#[cfg(feature = "wgpu")]` / `#[cfg(feature = "cuda")]` gates — exactly as shown for the existing groups. All four new groups must be added to `register_all` so every compiled backend exercises them automatically.
- The `training` feature gates the entire bench. The bench requires `--features training` at minimum. The `atari_cnn` module is under `thrust_rl::policy::atari_cnn` and does not require an additional feature beyond `training`.

Existing constant section uses `SYNTH_BATCH = 64` for train_step groups and `DQN_BATCH = 64` for DQN. Both `PPOConfig::default()` and `DQNConfig::default()` use `batch_size = 64` in-tree. The CNN groups should introduce new batch-size constants (not reuse `SYNTH_BATCH`), e.g.:

```rust
/// Nature-DQN CNN bench batch sizes.
/// B=32 mirrors the Nature-DQN paper (Mnih 2015); B=256 probes
/// the GPU crossover point (Atari PPO rollout minibatch upper range).
/// In-tree trainer defaults are batch_size=64 for both DQN and PPO.
const CNN_BATCH_DQN: usize = 32;
const CNN_BATCH_PPO: usize = 256;
const CNN_N_ACTIONS: usize = 18; // standard ALE legal-action count
```

---

### New benchmark groups

Add four new `bench_*` functions and extend `register_all` to call them.

**API reference** (from `src/policy/atari_cnn.rs` on `origin/main`):
- `NatureDqnBurnPolicy::<B>::with_config(n_actions: usize, config: NatureDqnConfig, device: &B::Device) -> Self`
  — `forward(&self, obs: Tensor<B, 4>) -> (Tensor<B, 2>, Tensor<B, 2>)` → (logits [B,A], values [B,1])
  — `evaluate_actions(&self, obs: Tensor<B, 4>, actions: Tensor<B, 1, Int>) -> (Tensor<B, 1>, Tensor<B, 1>, Tensor<B, 1>)` → (log_probs, entropy, values)
- `NatureDqnQNetwork::<B>::with_config(n_actions: usize, config: NatureDqnConfig, device: &B::Device) -> Self`
  — `forward(&self, obs: Tensor<B, 4>) -> Tensor<B, 2>` → Q-values [B,A]
- `NatureDqnConfig::default().with_seed(42)` — seeded config

Import: `use thrust_rl::policy::atari_cnn::{NatureDqnBurnPolicy, NatureDqnConfig, NatureDqnQNetwork};`

**Note**: `forward` takes `Tensor<B, 4>` in NCHW layout directly — no flat-reshape needed in these benches (the reshape-in-closure pattern is only needed when the rollout buffer hands flat [B, C*H*W] tensors to trainers).

#### Group 1 — `nature_dqn_policy_forward/{suffix}`

Measures isolated **inference** cost of `NatureDqnBurnPolicy::forward` (no autograd). Two `bench_function` entries: `"b32"` and `"b256"`.

Setup closure (untimed): construct policy with `AutodiffBackend` (so `register_all` signature is satisfied), build seeded synthetic NCHW input at the target batch size.

Timed closure: call `policy.forward(obs.clone())`, `black_box((logits, values))`.

Synthetic input construction pattern:
```rust
let data: Vec<f32> = (0..batch * 4 * 84 * 84)
    .map(|i| ((i as f32 * 0.001).sin().abs()))
    .collect();
let obs = Tensor::<B, 4>::from_data(
    TensorData::new(data, [batch, 4, 84, 84]),
    device,
);
```
Use `Throughput::Elements(batch as u64)`.

#### Group 2 — `nature_dqn_qnet_forward/{suffix}`

Same structure but for `NatureDqnQNetwork::forward`. Two bench_function entries: `"b32"`, `"b256"`. `black_box(q_values)`.

#### Group 3 — `nature_dqn_policy_train_step/{suffix}`

Measures **forward + backward + Adam optimizer step** for `NatureDqnBurnPolicy` — the training cost the PPO trainer would pay per minibatch update on Atari observations. Two entries: `"b32"`, `"b256"`.

Setup closure (untimed): construct policy, build `BurnOptimizer<B, NatureDqnBurnPolicy<B>, _>` wrapping `AdamConfig::new().init()` at lr 2.5e-4, build obs tensor + action tensor.

Timed closure (pseudo-code):
```rust
let (logits, values) = policy.forward(obs.clone());
// dummy actor-critic loss: log-softmax cross-entropy + value MSE
let loss = logits.log_softmax(1)
    .select(1, actions.clone())
    .mean()
    .neg()
    + values.powf_scalar(2.0).mean();
let grads = GradientsParams::from_grads(loss.backward(), &policy);
policy = optimizer.step(2.5e-4, policy, grads);
black_box(&policy);
```

Use the Burn `GradientsParams` + `optimizer.step(lr, module, grads)` idiom as used throughout the existing trainers. `Throughput::Elements(batch as u64)`.

#### Group 4 — `nature_dqn_qnet_train_step/{suffix}`

Same structure for `NatureDqnQNetwork` — forward + backward + Adam step. Analogous to the existing `dqn_train_step` group but using the CNN Q-network instead of `QNetworkBurn` (MLP). Dummy loss: `q_values.powf_scalar(2.0).mean()` (MSE against zero — sufficient to exercise the full backward graph). Two entries: `"b32"`, `"b256"`.

#### Extension to `register_all`

```rust
fn register_all<B: AutodiffBackend>(c: &mut Criterion, device: &B::Device, suffix: &str) {
    // existing eight groups ...
    bench_nature_dqn_policy_forward::<B>(c, device, suffix);
    bench_nature_dqn_qnet_forward::<B>(c, device, suffix);
    bench_nature_dqn_policy_train_step::<B>(c, device, suffix);
    bench_nature_dqn_qnet_train_step::<B>(c, device, suffix);
}
```

---

### How to run per backend

```bash
# CPU only (ndarray baseline — compiles everywhere, CI-safe).
cargo bench --features training --bench trainer_throughput \
    -- nature_dqn

# wgpu (Metal on this macOS arm64 host, Vulkan on Linux).
# Emits BOTH ndarray and wgpu groups in one run.
cargo bench --features "training,wgpu" --bench trainer_throughput \
    -- nature_dqn

# Quick smoke (short warm-up/measurement windows):
cargo bench --features "training,wgpu" --bench trainer_throughput \
    -- --warm-up-time 1 --measurement-time 3 nature_dqn

# cuda (Linux + NVIDIA + CUDA toolkit — run on alc-2, see below).
cargo bench --features "training,cuda" --bench trainer_throughput \
    --warm-up-time 2 --measurement-time 10 -- nature_dqn
```

The `-- nature_dqn` filter runs only the new CNN groups; omit it to run the full suite.

---

### `docs/BURN_BACKENDS.md` update contract

The current doc structure relevant to this issue:

1. **"Throughput benchmarking on GPU backends" → "Reading the results"** — a table mapping logical group names to `/{ndarray}`, `/{wgpu}`, `/{cuda}` column names.
   **Required change**: extend the table with four new rows for the CNN groups:

   | Logical group | CPU group id | wgpu group id | cuda group id |
   |---|---|---|---|
   | Nature-DQN policy forward | `nature_dqn_policy_forward/ndarray` | `…/wgpu` | `…/cuda` |
   | Nature-DQN Q-net forward | `nature_dqn_qnet_forward/ndarray` | `…/wgpu` | `…/cuda` |
   | Nature-DQN policy train step | `nature_dqn_policy_train_step/ndarray` | `…/wgpu` | `…/cuda` |
   | Nature-DQN Q-net train step | `nature_dqn_qnet_train_step/ndarray` | `…/wgpu` | `…/cuda` |

2. **After the existing "cuda column (issue #219, measured 2026-06-28)" section** — add a new subsection:

   ```
   ### Measured large-net (Nature-DQN CNN) throughput (issue #328)
   ```

   This section must contain:
   - Host setup block: macOS arm64, Apple M3 Ultra, Rust stable, Burn 0.21
   - The run command used
   - A results table with columns: Benchmark group | NdArray (CPU) | wgpu/Metal | cuda
   - CPU and wgpu columns filled in by the builder (measured locally on this host)
   - The cuda column marked with a clearly-labelled placeholder:

     ```
     | `nature_dqn_policy_train_step` (b256) | X ms | Y ms | **cuda: pending — run on alc-2** |
     ```

   - A footnote: "**cuda: pending operator run on alc-2.** Command: `cargo bench --features \"training,cuda\" --bench trainer_throughput --warm-up-time 2 --measurement-time 10 -- nature_dqn`. Host: `sphere@alc-2` over Tailscale (Ubuntu 24.04, RTX 4090, CUDA 12.x). See the existing cuda column above for environment setup reference."

3. **"Interpretation" / crossover verdict** — the current Interpretation section ends with the small-net finding ("CPU NdArray wins every group by 4.4–9.5×") and names CNN policies as what would flip the conclusion. The update must add a paragraph immediately after the existing interpretation block (or as a subsection titled **"Large-net verdict (Nature-DQN CNN)"**) that:
   - States the outcome: does the CNN workload flip the CPU-wins conclusion at B=32 and/or B=256?
   - References the specific groups where crossover does or does not occur.
   - Provides a clear recommendation update: if GPU wins at B≥256, update the "Picking the right backend" table's guidance for image-env training.
   - If cuda numbers are pending, states: "The large-net wgpu/Metal numbers above are available; the cuda crossover verdict is pending the alc-2 operator run."

---

### CUDA operator boundary

The builder MUST NOT block on CUDA access. The correct deliverable split is:

- **Builder lands**: benchmark code + ndarray (CPU) numbers + wgpu/Metal numbers (measurable on this macOS arm64 M3 Ultra host) + clearly-marked `cuda: pending` placeholders in the doc table with the exact command above.
- **Operator follow-up**: runs the cuda bench on `sphere@alc-2` (Tailscale), fills in the cuda column, removes the placeholder, and updates the crossover verdict if the cuda result changes the conclusion.
- The PR description must call this out explicitly: "cuda column in BURN_BACKENDS.md is a placeholder pending operator run on alc-2."

---

### Scope guard

This issue is **benches + docs only**. Strictly out of scope:

- No changes to `src/policy/atari_cnn.rs` or any other `src/` file.
- No env work (no ALE/subprocess integration).
- No trainer changes.
- No new examples.
- No changes to `Cargo.toml` (all required features already exist: `training`, `wgpu`, `cuda`).

If the builder discovers a bug in `atari_cnn.rs` during bench construction, file a separate issue rather than fixing it in-scope.

---

### Affected Files

- `benches/trainer_throughput.rs` — add four new `bench_nature_dqn_*` functions + extend `register_all` + add `CNN_BATCH_DQN`, `CNN_BATCH_PPO`, `CNN_N_ACTIONS` constants + extend module-level doc comment
- `docs/BURN_BACKENDS.md` — extend "Reading the results" table + add "Measured large-net CNN throughput" subsection + extend/add crossover verdict paragraph

Both paths confirmed present on `origin/main`.

---

### Acceptance Criteria (refined)

- [ ] `benches/trainer_throughput.rs` has four new groups: `nature_dqn_policy_forward`, `nature_dqn_qnet_forward`, `nature_dqn_policy_train_step`, `nature_dqn_qnet_train_step`, each with `"b32"` and `"b256"` bench_function entries.
- [ ] All four groups are registered in `register_all` and are exercised by the `wgpu` and `cuda` feature gates without any logic duplication.
- [ ] `cargo bench --features training --bench trainer_throughput -- nature_dqn` runs without panic on a CPU-only host.
- [ ] `cargo bench --features "training,wgpu" --bench trainer_throughput -- nature_dqn` runs without panic on this macOS arm64 host.
- [ ] `docs/BURN_BACKENDS.md` "Reading the results" table is extended with the four CNN group rows.
- [ ] `docs/BURN_BACKENDS.md` has a "Measured large-net (Nature-DQN CNN) throughput" subsection with measured CPU/wgpu numbers and a clearly-labelled `cuda: pending` placeholder containing the exact alc-2 run command.
- [ ] The crossover verdict paragraph states the conclusion (or notes it is partial pending cuda).
- [ ] `cargo test --features training`, `cargo clippy --features training -- -D warnings`, `cargo fmt --check`, and `cargo doc --features training --no-deps` all pass.
- [ ] No changes to `src/`, `Cargo.toml`, examples, or trainer code.

---

### Test Plan

- [ ] Manual verification: `cargo bench --features training --bench trainer_throughput -- --warm-up-time 1 --measurement-time 3 nature_dqn` completes without panic and prints criterion output for all eight new groups (4 groups × 2 batch sizes, suffixed `/ndarray`).
- [ ] Manual verification (GPU): repeat with `--features "training,wgpu"` on this macOS arm64 host; both `/ndarray` and `/wgpu` groups appear.
- [ ] Automated tests: `cargo test --features training` passes (no regressions; no new test files required since this is a bench-only change).
- [ ] Regression guard: existing eight groups (`a2c_train_step`, `ppo_train_step`, etc.) still compile and run unchanged.
- [ ] Doc build: `cargo doc --features training --no-deps` builds without warnings.
- [ ] Edge case: `cargo bench --features training --bench trainer_throughput` (without filter) runs the full twelve-group suite without error.
